qt/aqt/browser.py

```diff
@@ -1132,15 +1142,44 @@ def _favTree(self, root) -> None:
             root.addChild(item)
 
     def _userTagTree(self, root) -> None:
-        assert self.col
-        for t in self.col.tags.all():
-            item = SidebarItem(
-                t,
-                ":/icons/tag.svg",
-                lambda t=t: self.setFilter("tag", t),  # type: ignore
-                item_type=SidebarItemType.TAG,
-            )
-            root.addChild(item)
+        tree = self.col.backend.tag_tree()
+
+        def fillGroups(root, nodes: Sequence[TagTreeNode], head=""):
+            for node in nodes:
+
+                def set_filter():
+                    full_name = head + node.name  # pylint: disable=cell-var-from-loop
+                    return lambda: self.setFilter("tag", full_name)
+
+                def toggle_expand():
+
+                    full_name = head + node.name  # pylint: disable=cell-var-from-loop
+
+                    def toggle(_):
+                        try:
+                            self.mw.col.tags.toggle_browser_collapse(full_name)
+                        except NotFoundError:
+                            # tag is missing, register it first
+                            self.mw.col.tags.register([full_name])
+                            self.mw.col.tags.toggle_browser_collapse(full_name)
+
```

Comment:
Instead of the backend exposing get() and update() methods and requiring the frontend to check for the missing case, perhaps it would be nicer to just expose a set_tag_collapsed(tag: str, collapsed: bool) routine, where the backend takes care of fetching, creating if necessary, and setting the value? For collapsed we should be able to pass in 'not node.collapsed' to get toggling behaviour.